pulp_container/app/registry_api.py

```diff
@@ -1001,26 +985,58 @@ def put(self, request, path, pk=None):
                 tag = models.Tag.objects.get(name=tag.name, tagged_manifest=manifest)
                 tag.touch()
 
+            add_content_units += [str(tag.pk), str(manifest.pk)]
+
+            if is_manifest_list:
+                if not pending_blobs:
+                    pending_blobs = repository.pending_blobs.values_list("pk")
+
+                blobs_qs = models.Blob.objects.filter(
+                    pk__in=pending_blobs.union(found_manifests.values_list("blobs__pk"))
+                )
+                # we need to add pending blobs referenced by pending manifests
+                add_content_units += [str(blob.pk) for blob in blobs_qs]
+
+                add_content_units += [str(manifest.pk) for manifest in found_manifests]
+
             tags_to_remove = models.Tag.objects.filter(
                 pk__in=repository.latest_version().content.all(), name=tag
             ).exclude(tagged_manifest=manifest)
-            add_content_units = [str(tag.pk), str(manifest.pk)]
             remove_content_units = [str(pk) for pk in tags_to_remove.values_list("pk")]
-        else:
-            add_content_units = [str(manifest.pk)]
-            remove_content_units = []
 
-        dispatched_task = dispatch(
-            add_and_remove,
-            exclusive_resources=[repository],
-            kwargs={
-                "repository_pk": str(repository.pk),
-                "add_content_units": add_content_units,
-                "remove_content_units": remove_content_units,
-            },
-        )
+            immediate_task = dispatch(
+                add_and_remove,
+                exclusive_resources=[repository],
+                kwargs={
+                    "repository_pk": str(repository.pk),
+                    "add_content_units": add_content_units,
+                    "remove_content_units": remove_content_units,
+                },
+                immediate=True,
+                deferred=True,
+            )
 
-        if has_task_completed(dispatched_task):
+            if immediate_task.state == "completed":
+                if found_manifests:
+                    repository.pending_manifests.remove(*found_manifests)
```

Comment:
if you make this task deferrable and it will not complete within the allocated 3 secs waiting time, it will then be passed to worker. who will make sure the pending content is removed from this list? we should probably first remove and then dispatch the task, otherwise once the task is picked up by worker and it completes there will be noone who'd remove pending content

Reply:
Good catch. I had not thought about that.

Reply:
In the end, it would be an orphan cleanup that will handle this situation. But, I still think removing it beforehand is much better.

Reply:
definitely, also because orphan clean up is not regularly running task and god who know when it will be triggered

Reply:
@mdellweg, do you suggest calling `repository.pending_[manifests|blobs].clear()` in the `finalize_new_version` hook? Can we rule out any problems with removing pending manifests/blobs when other users are pushing data to the same repository?

Reply:
No, of course we should only remove the ones we added in that repository version.
Concurrent pushes is definitely a thing to consider.